File: lib-esphome-client/src/main/java/io/github/arhor/esphome/client/async/model/EspHomeEntity.java
package io.github.arhor.esphome.client.async.model;

import java.util.List;

public sealed interface EspHomeEntity {

    int key();

    String objectId();

    String name();

    record BinarySensor(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String deviceClass,
        boolean isStatusBinarySensor,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Cover(
        int key,
        String objectId,
        String name,
        String uniqueId,
        boolean assumedState,
        boolean supportsPosition,
        boolean supportsTilt,
        String deviceClass,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory,
        boolean supportsStop
    ) implements EspHomeEntity {}

    record Fan(
        int key,
        String objectId,
        String name,
        String uniqueId,
        boolean supportsOscillation,
        boolean supportsSpeed,
        boolean supportsDirection,
        int supportedSpeedCount,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory,
        List<String> supportedPresetModes
    ) implements EspHomeEntity {}

    record Light(
        int key,
        String objectId,
        String name,
        String uniqueId,
        List<ColorMode> supportedColorModes,
        float minMireds,
        float maxMireds,
        List<String> effects,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Sensor(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        String unitOfMeasurement,
        int accuracyDecimals,
        boolean forceUpdate,
        String deviceClass,
        SensorStateClass stateClass,
        SensorLastResetType legacyLastResetType,
        boolean disabledByDefault,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Switch(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean assumedState,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass
    ) implements EspHomeEntity {}

    record TextSensor(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass
    ) implements EspHomeEntity {}

    record Service(
        int key,
        String objectId,
        String name,
        List<ServiceArgument> args
    ) implements EspHomeEntity {}

    record Camera(
        int key,
        String objectId,
        String name,
        String uniqueId,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Climate(
        int key,
        String objectId,
        String name,
        String uniqueId,
        boolean supportsCurrentTemperature,
        boolean supportsTwoPointTargetTemperature,
        List<ClimateMode> supportedModes,
        float visualMinTemperature,
        float visualMaxTemperature,
        float visualTargetTemperatureStep,
        boolean legacySupportsAway,
        boolean supportsAction,
        List<ClimateFanMode> supportedFanModes,
        List<ClimateSwingMode> supportedSwingModes,
        List<String> supportedCustomFanModes,
        List<ClimatePreset> supportedPresets,
        List<String> supportedCustomPresets,
        boolean disabledByDefault,
        String icon,
        EntityCategory entityCategory,
        float visualCurrentTemperatureStep,
        boolean supportsCurrentHumidity,
        boolean supportsTargetHumidity,
        float visualMinHumidity,
        float visualMaxHumidity
    ) implements EspHomeEntity {}

    record Number(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        float minValue,
        float maxValue,
        float step,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String unitOfMeasurement,
        NumberMode mode,
        String deviceClass
    ) implements EspHomeEntity {}

    record Select(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        List<String> options,
        boolean disabledByDefault,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Siren(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        List<String> tones,
        boolean supportsDuration,
        boolean supportsVolume,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Lock(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        boolean assumedState,
        boolean supportsOpen,
        boolean requiresCode,
        String codeFormat
    ) implements EspHomeEntity {}

    record Button(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass
    ) implements EspHomeEntity {}

    record MediaPlayer(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        boolean supportsPause,
        List<MediaPlayerFormat> supportedFormats
    ) implements EspHomeEntity {}

    record AlarmControlPanel(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        int supportedFeatures,
        boolean requiresCode,
        boolean requiresCodeToArm
    ) implements EspHomeEntity {}

    record Text(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        int minLength,
        int maxLength,
        String pattern,
        TextMode mode
    ) implements EspHomeEntity {}

    record Date(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Time(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Event(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass,
        List<String> eventTypes
    ) implements EspHomeEntity {}

    record Valve(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass,
        boolean assumedState,
        boolean supportsPosition,
        boolean supportsStop
    ) implements EspHomeEntity {}

    record DateTime(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory
    ) implements EspHomeEntity {}

    record Update(
        int key,
        String objectId,
        String name,
        String uniqueId,
        String icon,
        boolean disabledByDefault,
        EntityCategory entityCategory,
        String deviceClass
    ) implements EspHomeEntity {}
}

File: lib-esphome-client/src/main/kotlin/io/github/arhor/esphome/client/model/EspHomeEntity.kt
package io.github.arhor.esphome.client.model

import io.github.arhor.esphome.client.proto.ListEntitiesAlarmControlPanelResponse
import io.github.arhor.esphome.client.proto.ListEntitiesBinarySensorResponse
import io.github.arhor.esphome.client.proto.ListEntitiesButtonResponse
import io.github.arhor.esphome.client.proto.ListEntitiesCameraResponse
import io.github.arhor.esphome.client.proto.ListEntitiesClimateResponse
import io.github.arhor.esphome.client.proto.ListEntitiesCoverResponse
import io.github.arhor.esphome.client.proto.ListEntitiesDateResponse
import io.github.arhor.esphome.client.proto.ListEntitiesDateTimeResponse
import io.github.arhor.esphome.client.proto.ListEntitiesEventResponse
import io.github.arhor.esphome.client.proto.ListEntitiesFanResponse
import io.github.arhor.esphome.client.proto.ListEntitiesLightResponse
import io.github.arhor.esphome.client.proto.ListEntitiesLockResponse
import io.github.arhor.esphome.client.proto.ListEntitiesMediaPlayerResponse
import io.github.arhor.esphome.client.proto.ListEntitiesNumberResponse
import io.github.arhor.esphome.client.proto.ListEntitiesSelectResponse
import io.github.arhor.esphome.client.proto.ListEntitiesSensorResponse
import io.github.arhor.esphome.client.proto.ListEntitiesServicesResponse
import io.github.arhor.esphome.client.proto.ListEntitiesSirenResponse
import io.github.arhor.esphome.client.proto.ListEntitiesSwitchResponse
import io.github.arhor.esphome.client.proto.ListEntitiesTextResponse
import io.github.arhor.esphome.client.proto.ListEntitiesTextSensorResponse
import io.github.arhor.esphome.client.proto.ListEntitiesTimeResponse
import io.github.arhor.esphome.client.proto.ListEntitiesUpdateResponse
import io.github.arhor.esphome.client.proto.ListEntitiesValveResponse

sealed interface EspHomeEntity {
    val key: Int
    val objectId: String
    val name: String

    @JvmInline
    value class BinarySensor(val raw: ListEntitiesBinarySensorResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Cover(val raw: ListEntitiesCoverResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Fan(val raw: ListEntitiesFanResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Light(val raw: ListEntitiesLightResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Sensor(val raw: ListEntitiesSensorResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Switch(val raw: ListEntitiesSwitchResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class TextSensor(val raw: ListEntitiesTextSensorResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Service(val raw: ListEntitiesServicesResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.name

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Camera(val raw: ListEntitiesCameraResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Climate(val raw: ListEntitiesClimateResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Number(val raw: ListEntitiesNumberResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Select(val raw: ListEntitiesSelectResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Siren(val raw: ListEntitiesSirenResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Lock(val raw: ListEntitiesLockResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Button(val raw: ListEntitiesButtonResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class MediaPlayer(val raw: ListEntitiesMediaPlayerResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class AlarmControlPanel(val raw: ListEntitiesAlarmControlPanelResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Text(val raw: ListEntitiesTextResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Date(val raw: ListEntitiesDateResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Time(val raw: ListEntitiesTimeResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Event(val raw: ListEntitiesEventResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Valve(val raw: ListEntitiesValveResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class DateTime(val raw: ListEntitiesDateTimeResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }

    @JvmInline
    value class Update(val raw: ListEntitiesUpdateResponse) : EspHomeEntity {
        override val key: Int
            get() = raw.key

        override val objectId: String
            get() = raw.objectId

        override val name: String
            get() = raw.name
    }
}
